### PR TITLE
`Assessment`: Fix score view for accepted complaints

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
I recently noticed that after accepting a complaint, the scores page would not display the updated score of the student, but instead would stay on the first result before the complaint got accepted. This PR fixes that.

### Description
The wrong complaint score was introduced by #6410. While in the exam mode, all results (of all correction rounds are needed to allow the instructor to navigate into on of the specific results (of one selected correction round). This however leads to the server response now containing two elements in case of complaints: `[result before complaint, result after complaint]`. The server returns that as an unordered Set, the client then sorts the list to have the earlier results first in the array. The latest result (that should be displayed on the scores page) therefore is last in the array, and now `last()` gets used to determine the result to show.

When working with complaints, the "View the Complaint" button also was not working, it opened the result before accepting the complaint. This is now also fixed by incrementing the correction round when necessary.

We also now hide the "start new assessment" button for programming practice mode participations, here manual assessments should not be possible.

To simplify the code, these buttons got extracted into their own component.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with manual assessment enabled
- 1 Exercise of any other type

Open the scores page and ensure that
- starting a new assessment works
- canceling and continuing an ongoing assessment works
- viewing accepted and rejected complaints works (for accepted complaints, the new result is shown)
- no "start assessment" button is shown for practice mode submissions.

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise with manual assessment and 2 correction rounds, also an exercise of any other type.

Open the scores page and ensure that
- starting a new assessment works for correction rounds 1 and 2
- canceling and continuing an ongoing assessment works for both correction rounds
- viewing accepted and rejected complaints works (for accepted complaints, the new result is shown)
(see also testing steps of #6410 and issue #6395)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [x] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exercise-scores.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5717/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-scores/exercise-scores.component.ts.html) | 96.12% | ✅  |
| [manage-assessment-buttons.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5717/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-scores/manage-assessment-buttons.component.ts.html) | 66.66% | ✅  |

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts [hunk 1]
```diff
@@ -189,7 +171,11 @@ export class ExerciseScoresComponent implements OnInit, OnDestroy {
             participation.results?.forEach((result, index) => {
                 participation.results![index].durationInMinutes = dayjs(result.completionDate).diff(participation.initializationDate, 'seconds');
             });
+            // sort the results from old to new.
+            // the result of the first correction round will be at index 0,
+            // the result of a complaints or the second correction at index 1.
             participation.results?.sort((result1, result2) => (result1.id ?? 0) - (result2.id ?? 0));
+            participation.results?.last();
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:178]:
```suggestion
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Other than that the code looks good

**Comment by Ceiridge** [comment]
- Tested on test2 with @MaximilianJG 
- The score page didn't show anything about assessments or gradings
- The complaints were visible for the student after accepting/rejecting
- Canceling and continuing assessments works
- **But**: Practice mode exercises with submissions are not shown at all in the assessment dashboard and thus also the assessment cannot be started

image

image

**Comment by MaximilianJG** [comment]
Worked mainly as expected on test server 2. (Testing session with @Ceiridge) 

Student was able to create a complaint, which we could assess. This complaint was then sent to the student which he could then view. We were also able to save and continue / cancel assessments. 

Nonetheless we had one issue. When the student started a practice exercise submission this was not viewable in the assessment dashboard. (only the actual submission) 
Screenshot 2023-11-15 at 17 04 34

**Comment by Strohgelaender** [comment]
> But: Practice mode exercises with submissions are not shown at all in the assessment dashboard and thus also the assessment cannot be started

That's correct and expected. Practice mode submission are just for the student to try out changes, they are not assessable and should not show up here. The issue fixed in this PR was that on the scores page an additional "assess submission" was shown, even though assessment should not be allowed.

**Review by tobias-lippert — DISMISSED** [review]
Code LGTM

File: src/main/webapp/app/exercises/shared/exercise-scores/manage-assessment-buttons.component.html [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            <button
+                *ngIf="
+                    newManualResultAllowed &&
+                    participation.results?.[correctionRound]?.assessmentType &&
+                    !participation.results?.[correctionRound]?.completionDate &&
+                    participation.results?.[correctionRound]?.assessmentType !== AssessmentType.AUTOMATIC
+                "
+                (click)="cancelAssessment(participation.results?.[correctionRound]!, participation)"
+                [disabled]="isLoading"
+                class="btn btn-danger btn-sm mb-1"
```
Comment (maximiliansoelch) [inline src/main/webapp/app/exercises/shared/exercise-scores/manage-assessment-buttons.component.html:48]: Noticed that the _cancel assessment_ button is misaligned compared to the _continue assessment_ button.
It was not introduced by this PR, but should be fixed anyway:
image

```suggestion
                class="btn btn-sm me-1 mb-2 btn-danger"
```

**Review by maximiliansoelch — APPROVED** [review]
Code looks good.
Tested on legacy ts1 and works as described. 

Just found a minor styling issue that was not introduced in this PR but should be fixed either here or in a follow-up.

**Review by maximiliansoelch — APPROVED** [review]
Reapprove after style correction commit

**Pull request merged**